Provider crash when renaming AS3 partition

### Environment

 * TMOS/Bigip Version:	BIG-IP 14.1.4 Build 0.0.11 Final
 * Terraform Version: 1.0.7
 * Terraform bigip provider Version: 1.11.1 -> 1.13.0

### Summary

In our environment, we name AS3 partitions after the common name of the service, the site and some other details. If one of these changes, the partition name changes. When the partition name changes the BIGIP provider crashes.

```
module.lbaas-pool.bigip_as3.application-site1["site1"]: Still modifying... [id=tfeedev_original_partition_name, 40s elapsed]
╷
│ Error: Plugin did not respond
│
│   with module.lbaas-pool.bigip_as3.application-site1["site1"],
│   on ..\modules\terraform_module-bigip-lbaas-pool\bigip_as3.tf line 276, in resource "bigip_as3" "application-site1":
│  276: resource "bigip_as3"  "application-site1" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-bigip_v1.13.0.exe plugin:

panic: runtime error: index out of range [1] with length 1

goroutine 41 [running]:
github.com/f5devcentral/go-bigip.(*BigIP).DeleteAs3Bigip(0xc0002e40e0, 0xc000236000, 0x57, 0x1, 0xc00020b320, 0x1, 0x1)
        github.com/f5devcentral/go-bigip@v0.0.0-20220302174144-f3febbee09dd/as3bigip.go:165 +0xde9
github.com/F5Networks/terraform-provider-bigip/bigip.resourceBigipAs3Update(0xc00025ad20, 0x14da380, 0xc0002e40e0, 0x0, 0x0)
        github.com/F5Networks/terraform-provider-bigip/bigip/resource_bigip_as3.go:327 +0x89f
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc0005ce080, 0xc000440730, 0xc000086420, 0x14da380, 0xc0002e40e0, 0x13b0e01, 0xc00006f780, 0xc000505470)
        github.com/hashicorp/terraform-plugin-sdk@v1.1.0/helper/schema/resource.go:311 +0x293
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc0005ce880, 0xc00006f8e8, 0xc000440730, 0xc000086420, 0xc0005c9678, 0xc000006478, 0x13b2b00)
        github.com/hashicorp/terraform-plugin-sdk@v1.1.0/helper/schema/provider.go:294 +0xa5
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc0000961b0, 0x1677f70, 0xc0005040f0, 0xc00055a060, 0xc0000961b0, 0xc0005040f0, 0xc0003c1a50)
        github.com/hashicorp/terraform-plugin-sdk@v1.1.0/internal/helper/plugin/grpc_provider.go:885 +0x8c5
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x1499820, 0xc0000961b0, 0x1677f70, 0xc0005040f0, 0xc00055a000, 0x0, 0x1677f70, 0xc0005040f0, 0xc00031c000, 0x2776)
        github.com/hashicorp/terraform-plugin-sdk@v1.1.0/internal/tfplugin5/tfplugin5.pb.go:3189 +0x222
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00014a160, 0x1680ed8, 0xc000051e00, 0xc00050c300, 0xc000504480, 0x1bedde0, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.23.0/server.go:995 +0x482
google.golang.org/grpc.(*Server).handleStream(0xc00014a160, 0x1680ed8, 0xc000051e00, 0xc00050c300, 0x0)
        google.golang.org/grpc@v1.23.0/server.go:1275 +0xd4c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0005021a0, 0xc00014a160, 0x1680ed8, 0xc000051e00, 0xc00050c300)
        google.golang.org/grpc@v1.23.0/server.go:710 +0xb2
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.23.0/server.go:708 +0xa8

Error: The terraform-provider-bigip_v1.13.0.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Steps To Reproduce

Steps to reproduce the behavior:
- Deploy any valid AS3 to a bigip device with terraform apply
- Rename the partition in the Terraform AS3 definition
- Reapply with terraform apply (error produced)
- Reapply with terraform apply again (success)

This issue is occurring in v 1.11.1, 1.12.0, 1.13.0 (possibly others)

1. Provide terraform resource config which you are facing trouble along with the output of it.
N/A

2. To get to know more about the issue, provide terraform debug logs
N/A

3. To capture debug logs, export TF_LOG variable with debug ( export TF_LOG= DEBUG ) before 
  runnning terraform apply/plan
N/A
5. As3/DO json along with the resource config( for AS3/DO resource issues )
N/A

### Expected Behavior
I expect to be able to rename an AS3 partition in a single run and not encounter an error.

### Actual Behavior
Provider crashes with `panic: runtime error: index out of range [1] with length 1` when renaming partitions.

--- a/bigip/resource_bigip_as3.go
+++ b/bigip/resource_bigip_as3.go
@@ -231,7 +231,18 @@ func resourceBigipAs3Create(d *schema.ResourceData, meta interface{}) error {
 func resourceBigipAs3Read(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*bigip.BigIP)
 	log.Printf("[INFO] Reading As3 config")
-	name := d.Id()
+	var name string
+	var tList string
+
+	if d.Get("as3_json") != nil {
+		tList, _, _ = client.GetTenantList(d.Get("as3_json").(string))
+	}
+	if d.Id() != "" && tList != "" {
+		name = tList
+	} else {
+		name = d.Id()
+	}
+
 	applicationList := d.Get("application_list").(string)
 	log.Printf("[DEBUG] Tenants in AS3 get call : %s", name)
 	if name != "Common" {
@@ -268,7 +279,18 @@ func resourceBigipAs3Read(d *schema.ResourceData, meta interface{}) error {
 func resourceBigipAs3Exists(d *schema.ResourceData, meta interface{}) (bool, error) {
 	client := meta.(*bigip.BigIP)
 	log.Printf("[INFO] Checking if As3 config exists in BIGIP")
-	name := d.Id()
+	var name string
+	var tList string
+
+	if d.Get("as3_json") != nil {
+		tList, _, _ = client.GetTenantList(d.Get("as3_json").(string))
+	}
+
+	if d.Id() != "" && tList != "" {
+		name = tList
+	} else {
+		name = d.Id()
+	}
 	applicationList := d.Get("application_list").(string)
 	tenantFilter := d.Get("tenant_filter").(string)
 	if tenantFilter != "" {
@@ -314,14 +336,14 @@ func resourceBigipAs3Update(d *schema.ResourceData, meta interface{}) error {
 	m.Lock()
 	defer m.Unlock()
 	log.Printf("[INFO] Updating As3 Config :%s", as3Json)
-	name := d.Id()
 	tenantList, _, _ := client.GetTenantList(as3Json)
+	oldTenantList := d.Get("tenant_list").(string)
 	tenantFilter := d.Get("tenant_filter").(string)
 	if tenantFilter == "" {
-		if tenantList != name {
+		if tenantList != oldTenantList {
 			_ = d.Set("tenant_list", tenantList)
 			newList := strings.Split(tenantList, ",")
-			oldList := strings.Split(name, ",")
+			oldList := strings.Split(oldTenantList, ",")
 			deletedTenants := client.TenantDifference(oldList, newList)
 			if deletedTenants != "" {
 				err, _ := client.DeleteAs3Bigip(deletedTenants)
@@ -363,7 +385,18 @@ func resourceBigipAs3Delete(d *schema.ResourceData, meta interface{}) error {
 	m.Lock()
 	defer m.Unlock()
 	log.Printf("[INFO] Deleting As3 config")
-	name := d.Id()
+	var name string
+	var tList string
+
+	if d.Get("as3_json") != nil {
+		tList, _, _ = client.GetTenantList(d.Get("as3_json").(string))
+	}
+
+	if d.Id() != "" && tList != "" {
+		name = tList
+	} else {
+		name = d.Id()
+	}
 	err, failedTenants := client.DeleteAs3Bigip(name)
 	if err != nil {
 		log.Printf("[ERROR] Unable to Delete: %v :", err)

--- a/bigip/resource_bigip_as3.go
+++ b/bigip/resource_bigip_as3.go
@@ -23,6 +23,7 @@ import (
 
 var x = 0
 var m sync.Mutex
+var createdTenants string
 
 func resourceBigipAs3() *schema.Resource {
 	return &schema.Resource{
@@ -225,6 +226,7 @@ func resourceBigipAs3Create(d *schema.ResourceData, meta interface{}) error {
 	} else {
 		d.SetId("Common")
 	}
+	createdTenants = d.Get("tenant_list").(string)
 	x++
 	return resourceBigipAs3Read(d, meta)
 }
@@ -233,9 +235,13 @@ func resourceBigipAs3Read(d *schema.ResourceData, meta interface{}) error {
 	log.Printf("[INFO] Reading As3 config")
 	var name string
 	var tList string
+	as3Json := d.Get("as3_json").(string)
 
 	if d.Get("as3_json") != nil {
-		tList, _, _ = client.GetTenantList(d.Get("as3_json").(string))
+		tList, _, _ = client.GetTenantList(as3Json)
+		if createdTenants != "" && createdTenants != tList {
+			tList = createdTenants
+		}
 	}
 	if d.Id() != "" && tList != "" {
 		name = tList
@@ -375,6 +381,7 @@ func resourceBigipAs3Update(d *schema.ResourceData, meta interface{}) error {
 			log.Printf("%v", err)
 		}
 	}
+	createdTenants = d.Get("tenant_list").(string)
 	_ = d.Set("task_id", taskID)
 	x++
 	return resourceBigipAs3Read(d, meta)

--- a/bigip/resource_bigip_as3_test.go
+++ b/bigip/resource_bigip_as3_test.go
@@ -71,6 +71,18 @@ resource "bigip_as3"  "as3-example" {
 }
 `
 
+var TestAs3Resourcegithub601a = `
+resource "bigip_as3"  "as3-example" {
+	as3_json = "${file("` + dir + `/../examples/as3/github601_a.json")}"
+}
+`
+
+var TestAs3Resourcegithub601b = `
+resource "bigip_as3"  "as3-example" {
+	as3_json = "${file("` + dir + `/../examples/as3/github601_b.json")}"
+}
+`
+
 func TestAccBigipAs3_create_SingleTenant(t *testing.T) {
 	resource.Test(t, resource.TestCase{
 		PreCheck: func() {
@@ -253,6 +265,31 @@ func TestAccBigipAs3Issue600(t *testing.T) {
 				Config: TestAs3Resourcegithub600,
 				Check: resource.ComposeTestCheckFunc(
 					testCheckAs3Exists("A1", true),
+				),
+			},
+		},
+	})
+}
+
+func TestAccBigipAs3Issue601(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAcctPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAs3Destroy,
+		Steps: []resource.TestStep{
+			{
+				Config: TestAs3Resourcegithub601a,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAs3Exists("Sample_01", true),
+				),
+			},
+			{
+				Config: TestAs3Resourcegithub601b,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAs3Exists("Sample_02", true),
+					testCheckAs3Exists("Sample_01", false),
 				),
 			},
 		},

--- a/examples/as3/as3_example.tmpl
+++ b/examples/as3/as3_example.tmpl
@@ -1,43 +1,43 @@
- {
-     "class": "AS3",
-     "action": "deploy",
-     "persist": true,
-     "declaration": {
-         "class": "ADC",
-         "schemaVersion": "3.0.0",
-         "id": "example-declaration-01",
-         "label": "Sample 1",
-         "remark": "Simple HTTP application with round robin pool",
-         ${tenant_name}: {
-             "class": "Tenant",
-             "defaultRouteDomain": 0,
-             "Application_1": {
-                 "class": "Application",
-                 "template": "http",
-             "serviceMain": {
-                 "class": "Service_HTTP",
-                 "virtualAddresses": [
-                     "10.0.2.10"
-                 ],
-                 "pool": "web_pool"
-                 },
-                 "web_pool": {
-                     "class": "Pool",
-                     "monitors": [
-                         "http"
-                     ],
-                     "members": [
-                         {
-                             "servicePort": 80,
-                             "serverAddresses": [
-                                 "192.0.1.100",
-                                 "192.0.1.110"
-                             ]
-                         }
-                     ]
-                 }
-             }
-         }
-     }
- }
+{
+    "class": "AS3",
+    "action": "deploy",
+    "persist": true,
+    "declaration": {
+        "class": "ADC",
+        "schemaVersion": "3.0.0",
+        "id": "example-declaration-01",
+        "label": "Sample 1",
+        "remark": "Simple HTTP application with round robin pool",
+        ${tenant_name}: {
+            "class": "Tenant",
+            "defaultRouteDomain": 0,
+            "Application_1": {
+                "class": "Application",
+                "template": "http",
+            "serviceMain": {
+                "class": "Service_HTTP",
+                "virtualAddresses": [
+                    "10.0.2.10"
+                ],
+                "pool": "web_pool"
+                },
+                "web_pool": {
+                    "class": "Pool",
+                    "monitors": [
+                        "http"
+                    ],
+                    "members": [
+                        {
+                            "servicePort": 80,
+                            "serverAddresses": [
+                                "192.0.1.100",
+                                "192.0.1.110"
+                            ]
+                        }
+                    ]
+                }
+            }
+        }
+    }
+}
 

